### ISSUE TITLE
Implement in place sorting of OrderMap and OrderSet

### ISSUE THREAD

**Issue by vitiral** [bluss/ordermap#55]
Feature request: Add in-place sorting methods to OrderMap (and OrderSet). Similar to the existing `.sorted_by`, these methods can allow the user to sort by either keys or values or both by supplying all of those to the closure.

For example:

```rust
fn sort_by<F>(&mut self, cmp: F)
    where F: FnMut(&K, &V, &K, &V) -> Ordering;
```

*Original issue:*

---

Feature request: add `sort_keys<K:Ord+Hash>(&mut self)` to do an inplace sort.

This library is INCREDIBLE when assertions need to be made because I can do an inplace sort and the diffs look good (when combined with `pretty_assertions` of course).

I did a naive implementation like this:
```
pub fn sort_ordermap<K:Ord+Hash, V>(m: &mut OrderMap<K, V>) {
    let mut ordered: Vec<_> = m.drain(..).collect();
    ordered.sort_by(|left, right| left.0.cmp(&right.0));
    m.extend(ordered.drain(..));
}

pub fn sort_orderset<K:Ord+Hash>(m: &mut OrderSet<K>) {
    let mut ordered: Vec<_> = m.drain(..).collect();
    ordered.sort_by(|left, right| left.cmp(&right));
    m.extend(ordered.drain(..));
}
```

I'm sure a better one can be done with full access to the underlying data structure.

**Comment by clarcharr** [comment]
You could also use `sort_by_key(|elem| elem.0)` for the `map` and just `sort` for the `orderset`.

**Comment by vitiral** [comment]
@clarcharr do you mean it could be named that? I suggested `sort` for OrderSet in the title of the issue. I'm not sure what `sort_by_key` means -- you *could* implement `sort_by` in addition to `sort_keys` -- both could be useful!

**Comment by vitiral** [comment]
@clarcharr do you mean it could be named that? I suggested `sort` for OrderSet in the title of the issue. I'm not sure what `sort_by_key` means -- you *could* implement `sort_by` in addition to `sort_keys` -- both could be useful!

**Comment by clarcharr** [comment]
Sorry; to clarify, you could replace the code in the post with:

```
pub fn sort_ordermap<K:Ord+Hash, V>(m: &mut OrderMap<K, V>) {
    let mut ordered: Vec<_> = m.drain(..).collect();
    ordered.sort_by_key(|x| x.0);
    m.extend(ordered);
}

pub fn sort_orderset<K:Ord+Hash>(m: &mut OrderSet<K>) {
    let mut ordered: Vec<_> = m.drain(..).collect();
    ordered.sort();
    m.extend(ordered);
}
```

**Comment by vitiral** [comment]
Oh, I didn't realize that `sort_by_key` was a thing! Thanks!

**Comment by bluss** [comment]
I agree with this feature request. No comment on whether it is possible easily

**Comment by bluss** [comment]
A side note about sort_by_key, is that the key version sometimes will need to clone the keys when the _by version does not.

**Comment by bluss** [comment]
I have taken the liberty to edit the original post to better reflect what I think we are seeking

**Comment by vitiral** [comment]
`sort_by` would be a good first start. Additional methods (like `sort_keys`) can be added later at other's requests.

**Comment by bluss** [comment]
Implemented as an experiment in #57, (building upon the retain in place PR, so that one needs to merge first).

Yes, instead of Vec's `.sort()` ordermap should probably have `.sort_keys()`, that sounds nice (And no `.sort()`)

**Comment by bluss** [comment]
Fixed in #57
